freeipa/src/main/java/com/sequenceiq/freeipa/flow/instance/reboot/action/AbstractRebootAction.java

```diff
@@ -0,0 +1,71 @@
+import static com.sequenceiq.cloudbreak.cloud.model.Location.location;
+import static com.sequenceiq.cloudbreak.cloud.model.Region.region;
+
+import java.util.Optional;
+
+import javax.inject.Inject;
+
+import org.springframework.statemachine.StateContext;
+
+import com.sequenceiq.cloudbreak.cloud.context.CloudContext;
+import com.sequenceiq.cloudbreak.cloud.model.CloudCredential;
+import com.sequenceiq.cloudbreak.cloud.model.Location;
+import com.sequenceiq.cloudbreak.common.event.Payload;
+import com.sequenceiq.cloudbreak.logger.MDCBuilder;
+import com.sequenceiq.flow.core.FlowParameters;
+import com.sequenceiq.freeipa.converter.cloud.CredentialToCloudCredentialConverter;
+import com.sequenceiq.freeipa.dto.Credential;
+import com.sequenceiq.freeipa.entity.InstanceMetaData;
+import com.sequenceiq.freeipa.entity.Stack;
+import com.sequenceiq.freeipa.flow.instance.InstanceFailureEvent;
+import com.sequenceiq.freeipa.flow.instance.reboot.RebootContext;
+import com.sequenceiq.freeipa.flow.instance.reboot.RebootEvent;
+import com.sequenceiq.freeipa.flow.instance.reboot.RebootState;
+import com.sequenceiq.freeipa.flow.stack.AbstractStackAction;
+import com.sequenceiq.freeipa.service.CredentialService;
+import com.sequenceiq.freeipa.service.stack.StackService;
+import com.sequenceiq.freeipa.service.stack.instance.InstanceMetaDataService;
+
+public abstract class AbstractRebootAction<P extends Payload>
+        extends AbstractStackAction<RebootState, RebootEvent, RebootContext, P>  {
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private InstanceMetaDataService instanceMetaDataService;
+
+    @Inject
+    private CredentialToCloudCredentialConverter credentialConverter;
+
+    @Inject
+    private CredentialService credentialService;
+
+    protected AbstractRebootAction(Class<P> payloadClass) {
+        super(payloadClass);
+    }
+
+    @Override
+    protected RebootContext createFlowContext(FlowParameters flowParameters, StateContext<RebootState, RebootEvent> stateContext, P payload) {
+        Long stackId = payload.getResourceId();
+        Stack stack = stackService.getByIdWithListsInTransaction(stackId);
+        MDCBuilder.buildMdcContext(stack);
+
+        Optional<InstanceMetaData> instance = instanceMetaDataService.getById(payload.getResourceId());
+        Location location = location(region(stack.getRegion()), availabilityZone(stack.getAvailabilityZone()));
+        CloudContext cloudContext = new CloudContext(stack.getId(), stack.getName(), stack.getCloudPlatform(), stack.getCloudPlatform(),
+                location, stack.getOwner(), stack.getAccountId());
+        Credential credential = credentialService.getCredentialByEnvCrn(stack.getEnvironmentCrn());
+        CloudCredential cloudCredential = credentialConverter.convert(credential);
+
+        return new RebootContext(flowParameters, stack, instance.get(), cloudContext, cloudCredential);
```

Comment:
if the `instance.get()` failes here is pretty strange, I think we should fail earlier

Reply:
fixed